library/error.c

```diff
@@ -213,701 +213,502 @@
+    {.code = -(MBEDTLS_ERR_THREADING_FEATURE_UNAVAILABLE), .description="THREADING - The selected feature is not available"},
+    {.code = -(MBEDTLS_ERR_THREADING_BAD_INPUT_DATA), .description="THREADING - Bad input parameters to function"},
+    {.code = -(MBEDTLS_ERR_THREADING_MUTEX_ERROR), .description="THREADING - Locking / unlocking / free failed with error code"},
 #endif /* MBEDTLS_THREADING_C */
 
 #if defined(MBEDTLS_XTEA_C)
-    if( use_ret == -(MBEDTLS_ERR_XTEA_INVALID_INPUT_LENGTH) )
-        mbedtls_snprintf( buf, buflen, "XTEA - The data input has an invalid length" );
-    if( use_ret == -(MBEDTLS_ERR_XTEA_HW_ACCEL_FAILED) )
-        mbedtls_snprintf( buf, buflen, "XTEA - XTEA hardware accelerator failed" );
+    {.code = -(MBEDTLS_ERR_XTEA_INVALID_INPUT_LENGTH), .description="XTEA - The data input has an invalid length"},
+    {.code = -(MBEDTLS_ERR_XTEA_HW_ACCEL_FAILED), .description="XTEA - XTEA hardware accelerator failed"},
 #endif /* MBEDTLS_XTEA_C */
-    // END generated code
+};
+
+#define NUM_LOW_LEVEL_ERRORS  ( sizeof(low_level_errors)/sizeof(mbedtls_error_t) )
+
+const char * mbedtls_high_level_strerr( int error_code )
+{
+    size_t i;
+    const char *error_description = NULL;
+
+    for(i = 0; i < NUM_HIGH_LEVEL_ERRORS; i++ )
+    {
+        if( high_level_errors[i].code == error_code )
+        {
+            error_description = high_level_errors[i].description;
+            break;
+        }
+    }
+
+    return error_description;
+}
+
+const char * mbedtls_low_level_strerr( int error_code )
+{
+    size_t i;
+    const char *error_description = NULL;
+
+    for(i = 0; i < NUM_LOW_LEVEL_ERRORS; i++ )
+    {
+        if( low_level_errors[i].code == error_code )
+        {
+            error_description = low_level_errors[i].description;
+            break;
+        }
+    }
+
+    return error_description;
+}
 
-    if( strlen( buf ) != 0 )
+void mbedtls_strerror( int ret, char *buf, size_t buflen )
+{
+    size_t len;
+    int use_ret;
+    const char * high_level_error_description = NULL;
+    const char * low_level_error_description = NULL;
+
+    if( buflen == 0 )
         return;
 
-    mbedtls_snprintf( buf, buflen, "UNKNOWN ERROR CODE (%04X)", use_ret );
+    memset( buf, 0x00, buflen );
+
+    if( ret < 0 )
+        ret = -ret;
+
+    if( ret & 0xFF80 )
+    {
+        use_ret = ret & 0xFF80;
+
+        // Translate high level error code.
+        high_level_error_description = mbedtls_high_level_strerr(use_ret);
+
+        if( high_level_error_description == NULL )
+            mbedtls_snprintf( buf, buflen, "UNKNOWN ERROR CODE (%04X)", use_ret );
+        else
+            mbedtls_snprintf( buf, buflen, "%s", high_level_error_description );
+
+        // Early return in case of a fatal error - do not try to translate low
+        // level code.
+        if(use_ret == -(MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE))
```

Comment:
CI says this isn't always defined. This block needs to be guarded by `#if defined(MBEDTLS_SSL_TLS_C)`.

I think that if you pass `tests/scripts/test-ref-configs.pl` that will resolve all the failures on CI.

Reply:
Thank you for the suggestion. Added the ```#if defined``` guard and ```tests/scripts/test-ref-configs.pl``` pass now.